pkg/ruler/manager.go

```diff
@@ -141,33 +140,53 @@ func (r *DefaultMultiTenantManager) syncRulesToManager(ctx context.Context, user
 		return
 	}
 
+	if !update {
+		level.Debug(r.logger).Log("msg", "rules have not changed, skipping rule manager update", "user", user)
+		return
+	}
+
+	manager, err := r.getOrCreateManager(ctx, user)
+	if err != nil {
+		r.lastReloadSuccessful.WithLabelValues(user).Set(0)
+		level.Error(r.logger).Log("msg", "unable to create rule manager", "user", user, "err", err)
+		return
+	}
+
+	level.Debug(r.logger).Log("msg", "updating rules", "user", user)
+	r.configUpdatesTotal.WithLabelValues(user).Inc()
+
+	err = manager.Update(r.cfg.EvaluationInterval, files, nil, r.cfg.ExternalURL.String(), nil)
+	if err != nil {
+		r.lastReloadSuccessful.WithLabelValues(user).Set(0)
+		level.Error(r.logger).Log("msg", "unable to update rule manager", "user", user, "err", err)
+		return
+	}
+
+	r.lastReloadSuccessful.WithLabelValues(user).Set(1)
+	r.lastReloadSuccessfulTimestamp.WithLabelValues(user).SetToCurrentTime()
+}
+
+// getOrCreateManager retrieves the user manager. if it doesn't exist, it will create and start it first
+func (r *DefaultMultiTenantManager) getOrCreateManager(ctx context.Context, user string) (RulesManager, error) {
+	r.userManagerMtx.Lock()
```

Comment:
We can do something slightly better here. Since we expect most managers to already exist, we could take a read lock first to check if it already exists. If not, we take the exclusive lock, we recheck and, if still doesn't exist, we create it.